### PR TITLE
FMT: add line break after opening brace in struct blocks

### PR DESCRIPTION
**Pull request by Kobzol**
This PR forces a line break after a left opening brace in struct blocks. This should bring struct block formatting closer to rustfmt. This change will probably ruin more tests, let's see what CI says.

Fixes: https://github.com/intellij-rust/intellij-rust/issues/5282

### CODE REVIEW

**Comment by Kobzol** [comment]
There is a regression with regard to `rustfmt` - the field block should be multiline only in structs, not in struct enum variants - currently my change applies to both of them. Is there a way to do this declaratively or do I have to use `Block.computeSpacing`?

**Comment by vlad20012** [comment]
Looks like `Block.computeSpacing` is the only way, To be honest, I'm not familiar with the formatter and it is mostly unmaintained.

**Comment by Kobzol** [comment]
Well, I tried something like
```kotlin
elementType1 == LBRACE && child2.node?.treeParent?.elementType == BLOCK_FIELDS && psi2.parent?.parent is RsStructItem ->
       return createSpacing(1, 1, 1, true, 0)
```
in `computeSpacing` and it sort of works, but it breaks some subtle conditions (like trailing commas), because the default spacing builder is not used if I override it here. As a last effort, I tried to reuse the output from the spacing builder in `computeSpacing` and just override it's newline count, but even then it still breaks for trailing commas.
```kotlin
val spacing = ctx.spacingBuilder.getSpacing(block, child1, child2)
  when (spacing) {
    is SpacingImpl -> return createSpacing(spacing.minSpaces, spacing.maxSpaces, 1, spacing.shouldKeepLineFeeds(), spacing.keepBlankLines)
    is DependantSpacingImpl -> return createSpacing(spacing.minSpaces, spacing.maxSpaces, 1, spacing.shouldKeepLineFeeds(), spacing.keepBlankLines)
  }
```

I didn't find how I could specify this grandparent relationship in the spacing builder. If it's not possible to do this cleanly, then maybe we should scratch this. Enabling it for enums would not play well with `rustfmt` (although the current implementation also formats differently than `rustfmt`).

**Comment by Kobzol** [comment]
It turned out that the default formatting was quite different from `rustfmt`, so I changed it slightly to match `rustfmt` more. Now trailing commas are also used for all struct block fields.

File: src/main/kotlin/org/rust/ide/formatter/processors/RsTrailingCommaFormatProcessor.kt [hunk 1]
```diff
@@ -84,7 +86,12 @@ fun CommaList.addTrailingCommaForElement(list: PsiElement): Boolean {
     )
     if (!trailingSpace.contains('\n')) return false
 
-    if (list.firstChild.getNextNonCommentSibling() == lastElement) return false
+    if (list.firstChild.getNextNonCommentSibling() == lastElement) {
+        // allow trailing comma only for struct block fields with a single item
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/formatter/processors/RsTrailingCommaFormatProcessor.kt:90]: What means "with a single item"? I don't get it :(

File: src/test/kotlin/org/rust/ide/refactoring/RsExtractEnumVariantTest.kt [hunk 2]
```diff
@@ -42,7 +42,11 @@ class RsExtractEnumVariantTest : RsTestBase() {
             V2
         }
     """, """
-        struct /*caret*/V1 { pub a: i32, pub b: bool, pub c: String }
+        struct /*caret*/V1 {
+            pub a: i32,
+            pub b: bool,
+            pub c: String
```
Comment (vlad20012) [inline src/test/kotlin/org/rust/ide/refactoring/RsExtractEnumVariantTest.kt:48]: Hmm, where is trailing comma here? Is it the same issue that `RsTrailingCommaFormatProcessor` doesn't invoke for some reason?

File: src/main/kotlin/org/rust/ide/formatter/processors/RsTrailingCommaFormatProcessor.kt (hunk 1 above)
Comment (Kobzol) [reply src/main/kotlin/org/rust/ide/formatter/processors/RsTrailingCommaFormatProcessor.kt:90]: Sorry for the confusion. `RsTrailingCommaFormatProcessor` does not add trailing commas to comma lists with a single element. So the plugin would format the following struct like this:
```rust
struct S {
    a: u32
}
```
while `rustfmt` formats it like this:
```rust
struct S {
    a: u32,
}
```
so I allowed the trailing comma for a comma list with a single element, but only in structs with named fields.

To make matters more complicated, the actual behaviour of `rustfmt` is this: it adds a trailing comma for fields only if the fields are multi-line. Fields are multi-line if:
1) They are inside struct with named fields ("classic" structs)
2) They are inside tuple struct or inside an enum variant and they are too long (but this happens rarely, the fields basically need to be longer than the maximum line length).

So this condition covers case 1), which is most common.

File: src/test/kotlin/org/rust/ide/refactoring/RsExtractEnumVariantTest.kt (hunk 2 above)
Comment (Kobzol) [reply src/test/kotlin/org/rust/ide/refactoring/RsExtractEnumVariantTest.kt:48]: Yes, I tried it now and the comma processor doesn't get invoked.

Comment (vlad20012) [reply src/test/kotlin/org/rust/ide/refactoring/RsExtractEnumVariantTest.kt:48]: Oh

**Comment by Kobzol** [comment]
@vlad20012 do you want to merge this PR? It should bring formatting closer to `rustfmt`, although it is not perfect.

**Comment by vlad20012** [comment]
It looks like `RsTrailingCommaFormatProcessor` is never called after intentions/quick fixes. I guess it's by design.

I am very sorry for such a long responce. For a long time, I was not familiar with the formatting subsystem, and I did not understand how it should work. Now I seem to understand more or less.

bors try

**Comment by Undin** [comment]
bors retry

**Pull request merged**